tutorials/plot_sleep.py

```diff
@@ -0,0 +1,275 @@
+raw_test.set_channel_types(mapping)
+events_test, _ = mne.events_from_annotations(
+    raw_test, event_id=annotation_desc_2_event_id, chunk_duration=30.)
+epochs_test = mne.Epochs(raw=raw_test, events=events_test, event_id=event_id,
+                         tmin=0., tmax=tmax, baseline=None)
+
+print(epochs_test)
+
+
+##############################################################################
+# .. _plot_sleep_extract_features:
+#
+# Feature Engineering
+# -------------------
+#
+# Observing the power spectrum density (PSD) plot of the :term:`epochs` grouped
+# by sleeping stage we can see that different sleep stages have different
+# signatures, and those signatures remain similar between Alice and Bob's data.
+#
+# The rest of this section we will create EEG features based on relative power
+# in specific frequency bands to discrete to capture this difference between
+# the sleep stages in our data.
+
+# visualize Alice vs. Bob PSD by sleep stage.
+fig, (ax1, ax2) = plt.subplots(ncols=2)
+
+# iterate over the subjects
+for ax, title, epochs in zip([ax1, ax2],
+                             ['Alice', 'Bob'],
+                             [epochs_train, epochs_test]):
+
+    for stage, color in zip(event_id.keys(), stage_colors):
+        epochs[stage].plot_psd(area_mode=None, color=color, ax=ax,
+                               fmin=0.1, fmax=20.)
+    ax.set_title(title)
+
+plt.legend(list(event_id.keys()))
+
+##############################################################################
+#
+# Design a sklearn compatible FunctionTransformer callable
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# We will now create a function to extract EEG features based on relative power
+# in specific frequency bands to be able to predict sleep stages from EEG
+# signals.
+
+def eeg_power_band(epochs):
+    """EEG relative power band feature extraction.
+
+    This function takes an ``mne.Epochs`` object and creates EEG features based
+    on relative power in specific frequency bands that are compatible with
+    scikit-learn.
+
+    Parameters
+    ----------
+    epochs : Epochs
+        The data.
+
+    Returns
+    -------
+    X : numpy array of shape [n_samples, 5]
+        Transformed data.
+    """
+    # specific frequency bands
+    FREQ_BANDS = {"delta": [0.5, 4.5],
+                  "theta": [4.5, 8.5],
+                  "alpha": [8.5, 11.5],
+                  "sigma": [11.5, 15.5],
+                  "beta": [15.5, 30]}
+
+    EEG_CHANNELS = ["EEG Fpz-Cz", "EEG Pz-Oz"]
+
+    sfreq = epochs.info['sfreq']
+    data = epochs.load_data().pick_channels(EEG_CHANNELS).get_data()
+    psds, freqs = psd_array_welch(data, sfreq, fmin=0.5, fmax=30.,
+                                  n_fft=512, n_overlap=256)
+    # Normalize the PSDs
+    psds /= np.sum(psds, axis=-1)[:, :, np.newaxis]
+
+    X = []
+    for _, (fmin, fmax) in FREQ_BANDS.items():
+        psds_band = psds[:, :, (freqs >= fmin) & (freqs < fmax)].mean(axis=-1)
```

Comment:
you could write

```
fmin >= freqs < fmax
```

Reply:
its a numpy array. You can do that with scalars, or single condition but not both. (or at least I don't know how to write it)

```py
In [5]: 1 <= 2 < 3                                                                                                                                                                                                 
Out[5]: True

In [6]: np.array(range(5)) < 3                                                                                                                                                                                     
Out[6]: array([ True,  True,  True, False, False])

In [7]: 1 <= np.array(range(5)) < 3                                                                                                                                                                                
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-b2202a63afa7> in <module>
----> 1 1 <= np.array(range(5)) < 3

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```